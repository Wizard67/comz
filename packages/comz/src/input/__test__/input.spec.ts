import { ref, nextTick } from 'vue'
import { mount } from '@vue/test-utils'
import Input from '../input.vue'

describe('Input.vue', () => {
  describe('props', () => {
    it('modelValue', () => {
      const value = 'input value'

      const wrapper = mount(Input, {
        props: {
          modelValue: value
        }
      })

      const element = wrapper.find('.cinput__field').element

      expect((element as HTMLInputElement).value).toBe(value)
      expect(wrapper.element).toMatchSnapshot()
    })

    it('placeholder', () => {
      const placeholder = 'text'

      const wrapper = mount(Input, {
        props: {
          modelValue: '',
          placeholder: placeholder
        }
      })

      const element = wrapper.get('.cinput__field').element

      expect((element as HTMLInputElement).placeholder).toBe(placeholder)
      expect(wrapper.element).toMatchSnapshot()
    })

    it('readonly', () => {
      const wrapper = mount(Input, {
        props: {
          modelValue: '',
          readonly: true
        }
      })

      expect(wrapper.classes()).toContain('cinput--readonly')
      expect(wrapper.element).toMatchSnapshot()
    })

    it('disabled', () => {
      const wrapper = mount(Input, {
        props: {
          modelValue: '',
          disabled: true
        }
      })

      expect(wrapper.classes()).toContain('cinput--disabled')
      expect(wrapper.element).toMatchSnapshot()
    })

    it('clearable', async () => {
      const wrapper = mount(Input, {
        props: {
          modelValue: '',
          clearable: true
        }
      })

      expect(wrapper.find('.cinput__clear').exists()).toBe(false)

      await wrapper.setProps({ modelValue: 'text' })
      expect(wrapper.find('.cinput__clear').exists()).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('events', () => {
    it('update:modelValue', async () => {
      const wrapper = mount(Input, {
        props: {
          modelValue: ''
        }
      })

      await wrapper.find('.cinput__field').setValue('hello')
      expect(wrapper.emitted()).toHaveProperty('update:modelValue')
      expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['hello'])
    })

    it('on-focus', () => {
      const wrapper = mount(Input, {
        attachTo: document.body,
        props: {
          modelValue: 'text'
        }
      })

      const element = wrapper.find('.cinput__field').element as HTMLInputElement
      element.focus()

      expect(wrapper.emitted()).toHaveProperty('on-focus')
      expect(wrapper.emitted()['on-focus'][0]).toEqual(['text'])
    })

    it('on-blur', () => {
      const wrapper = mount(Input, {
        attachTo: document.body,
        props: {
          modelValue: 'text'
        }
      })

      const element = wrapper.find('.cinput__field').element as HTMLInputElement
      element.focus()
      element.blur()

      expect(wrapper.emitted()).toHaveProperty('on-blur')
      expect(wrapper.emitted()['on-blur'][0]).toEqual(['text'])
    })

    it('on-clear', async () => {
      const value = ref('text')

      const wrapper = mount(Input, {
        attachTo: document.body,
        props: {
          modelValue: value,
          clearable: true
        }
      })

      const element = wrapper.find('.cinput__clear')
      element.trigger('click')
      expect(wrapper.emitted()).toHaveProperty('on-clear')

      value.value = ''
      await nextTick()

      const input = wrapper.find('.cinput__field').element as HTMLInputElement
      expect(input.value).toBe('')
      expect(wrapper.find('.cinput__clear').exists()).toBe(false)
    })
  })

  describe('slots', () => {
    it('prepend', () => {
      const content = "prepend"

      const wrapper = mount(Input, {
        props: {
          modelValue: ''
        },
        slots: {
          prepend: content
        }
      })
      
      const element = wrapper.find('.cinput__prepend')

      expect(element.exists()).toBe(true)
      expect(element.text()).toBe(content)
      expect(wrapper.element).toMatchSnapshot()
    })

    it('append', () => {
      const content = "append"

      const wrapper = mount(Input, {
        props: {
          modelValue: ''
        },
        slots: {
          append: content
        }
      })
      
      const element = wrapper.find('.cinput__append')

      expect(element.exists()).toBe(true)
      expect(element.text()).toBe(content)
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
