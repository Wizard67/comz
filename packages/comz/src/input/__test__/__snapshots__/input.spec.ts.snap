// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.vue props clearable 1`] = `
<div
  class="cinput"
>
  <!--v-if-->
  <div
    class="cinput__wrapper"
  >
    <input
      class="cinput__field"
      type="text"
    />
    <div
      class="cinput__clear"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM5.354 4.646a.5.5 0 1 0-.708.708L7.293 8l-2.647 2.646a.5.5 0 0 0 .708.708L8 8.707l2.646 2.647a.5.5 0 0 0 .708-.708L8.707 8l2.647-2.646a.5.5 0 0 0-.708-.708L8 7.293 5.354 4.646z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`Input.vue props disabled 1`] = `
<div
  class="cinput cinput--disabled"
>
  <!--v-if-->
  <div
    class="cinput__wrapper"
  >
    <input
      class="cinput__field cinput__field--disabled"
      disabled=""
      type="text"
    />
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`Input.vue props modelValue 1`] = `
<div
  class="cinput"
>
  <!--v-if-->
  <div
    class="cinput__wrapper"
  >
    <input
      class="cinput__field"
      type="text"
    />
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`Input.vue props placeholder 1`] = `
<div
  class="cinput"
>
  <!--v-if-->
  <div
    class="cinput__wrapper"
  >
    <input
      class="cinput__field"
      placeholder="text"
      type="text"
    />
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`Input.vue props readonly 1`] = `
<div
  class="cinput"
>
  <!--v-if-->
  <div
    class="cinput__wrapper"
  >
    <input
      class="cinput__field cinput__field--readonly"
      readonly=""
      type="text"
    />
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`Input.vue slots append 1`] = `
<div
  class="cinput"
>
  <!--v-if-->
  <div
    class="cinput__wrapper"
  >
    <input
      class="cinput__field"
      type="text"
    />
    <!--v-if-->
  </div>
  <div
    class="cinput__append"
  >
    
    
    append
    
    
  </div>
</div>
`;

exports[`Input.vue slots prepend 1`] = `
<div
  class="cinput"
>
  <div
    class="cinput__prepend"
  >
    
    
    prepend
    
    
  </div>
  <div
    class="cinput__wrapper"
  >
    <input
      class="cinput__field"
      type="text"
    />
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;
